_posts/2016-03-16-AngularTS.md

```diff
@@ -0,0 +1,413 @@
+In my custom services you will see examples of restangular in TypeScript.
+If you like to checkout what the difference is with $resource, you can check this [list](https://github.com/mgonto/restangular#differences-with-resource)
+
+{% highlight javascript %}
+/// <reference path="../../../typings/angularjs/angular.d.ts" />
+/// <reference path="../../../typings/restangular/restangular.d.ts" />
+
+
+module JWorks {
+
+   export interface IEmployeeService {
+       username:string;
+       employee:IEmployee;
+
+       getEmployee(): Employee;
+       setEmployee(employee:IEmployee);
+       getEmployeeByUsername(username:string):ng.IPromise<{}>;
+       getEmployeeByLink(href:string):ng.IPromise<{}>;
+   }
+
+   export class EmployeeService implements IEmployeeService {
+
+       static $inject = ["EmployeeRestangular", "$location"];
+
+       constructor(private employeeRestService:restangular.IService, private $location:ng.ILocationService, private employees:restangular.IElement,public employee:IEmployee,public username:string) {
+            username =window.sessionStorage.getItem("username");
+            this.employees = employeeRestService.all("employees");
+            employees.one(username).get().then((data:any)=> {
+                   this.setEmployee(data);
+               });
+           }
+       }
+              getEmployee():Employee {
+                  return this.employee;
+              }
+       
+              setEmployee(employee:IEmployee) {
+                  this.employee = employee;
+              }
+       
+              getEmployeeByUsername(username:string):ng.IPromise<{}> {
+                  return this.employees.one(username).get()
+       
+              }
+              getEmployeeByLink(href:string):ng.IPromise<{}>{
+                  return this.employees.oneUrl(href).get();
+              }       
+          }
+          angular.module("jworks360")
+              .service("EmployeeService", EmployeeService);       
+       }
+{% endhighlight %}
+
+In the above example, restangular has its own tsd file to reference to.
+For services it is best practice to declare an interface for data typing and get a clear view of the intent.
+The service class will implement all methods related to the data communication with the backend and returns a promise to the controllers or services that will inject this custom service.
+RestAngular has it’s own configuration you can modify in the .config component to point to the right api call.
+After the config you can inject the restangular service and use its services to build up a request to the backend.
+
+
+## Directives
+Custom directives allow you to create highly semantic and reusable components.
+A directive allows angular to manipulate the DOM and add it’s own behaviours. 
+These can be either a set of instructions or a JSON representation.
+Defining a directive in TypeScript we use the directive service angular provides. 
```

Comment:
Reword to:
**To define** a directive in TypeScript we use the directive service** that Angular** provides.